![Imgur](https://i.imgur.com/5pXzCIu.png)

# Data Science va Sun'iy Intellekt Praktikum

## 3-MODUL. MA’LUMOTLARGA ISHLOV BERISH

### 3.2 Ma'lumotlarni tozalash va tayyorlash

## Guruhlash. `.groupby()`
`.groupby()` meotdi yordamida jadvaldagi ma'lumotlarni biror ustun qiymatlari guruhlarga ajratib tahlil qilishimiz mumkin.

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("https://github.com/anvarnarz/praktikum_datasets/raw/main/automobile_data.csv", index_col=0)
df.head()

,company,body-style,wheel-base,length,engine-type,num-of-cylinders,horsepower,average-mileage,price
index,,,,,,,,,
0,alfa-romero,convertible,88.6,168.8,dohc,four,111,21,13495.0
1,alfa-romero,convertible,88.6,168.8,dohc,four,111,21,16500.0
2,alfa-romero,hatchback,94.5,171.2,ohcv,six,154,19,16500.0
3,audi,sedan,99.8,176.6,ohc,four,102,24,13950.0
4,audi,sedan,99.4,176.6,ohc,five,115,18,17450.0


Yuqoridagi jadvalda turli avtomobil brandlari haqida ma'lumotlar jamlangan. Keling boshlanishiga har bir brand uchun ustunlarning o'rtacha qiymatini topamiz:

In [9]:
df.groupby('company').mean(numeric_only=True)

,wheel-base,length,horsepower,average-mileage,price
company,,,,,
alfa-romero,90.566667,169.600000,125.333333,20.333333,15498.333333
audi,101.200000,180.800000,109.250000,20.000000,16392.500000
bmw,103.433333,185.033333,144.833333,19.000000,27213.333333
chevrolet,92.466667,151.933333,62.666667,41.000000,6007.000000
dodge,93.700000,157.300000,68.000000,31.000000,6303.000000
honda,96.500000,167.200000,92.333333,26.333333,10195.000000
isuzu,94.433333,160.833333,72.666667,33.333333,6785.000000
jaguar,109.333333,196.966667,204.666667,14.333333,34600.000000
mazda,95.900000,164.260000,75.400000,28.000000,9654.800000


Yuqoridagi natijani tahlil qilsak, `alfa-romero` mashinalarning o'rtacha narxi (`price`) `15498$`, `honda` mashinalarining o'rtacha ot kuchi (`horsepower`) `92.3` ga teng ekan.

Bitta ustun uchun qiymatlarni olish ham mumkin:

In [8]:
df.groupby('company').price.max() #max narxlarni chiqaramiz

,price
company,
alfa-romero,16500.0
audi,18920.0
bmw,41315.0
chevrolet,6575.0
dodge,6377.0
honda,12945.0
isuzu,6785.0
jaguar,36000.0
mazda,18344.0


Guruhlashda bir emas, bir nechta ustunlarni ham ko'rsatish mumkin:

In [11]:
df.groupby(['company','body-style']).mean(numeric_only=True)

wheel-base      length  horsepower  \
company       body-style                                        
alfa-romero   convertible   88.600000  168.800000  111.000000   
              hatchback     94.500000  171.200000  154.000000   
audi          sedan         99.666667  176.833333  109.000000   
              wagon        105.800000  192.700000  110.000000   
bmw           sedan        103.433333  185.033333  144.833333   
chevrolet     hatchback     91.450000  148.500000   59.000000   
              sedan         94.500000  158.800000   70.000000   
dodge         hatchback     93.700000  157.300000   68.000000   
honda         sedan         96.500000  172.250000  100.500000   
              wagon         96.500000  157.100000   76.000000   
isuzu         sedan         94.433333  160.833333   72.666667   
jaguar        sedan        109.333333  196.966667  204.666667   
mazda         hatchback     93.650000  161.575000   76.250000   
              sedan        104.900000  175.000000   72.000000   
mercedes-benz hardtop      112.000000  199.200000  184.000000   
              sedan        115.450000  199.500000  153.500000   
              wagon        110.000000  190.900000  123.000000   
mitsubishi    hatchback     93.700000  157.300000   68.000000   
              sedan         96.300000  172.400000   88.000000   
nissan        sedan         95.975000  170.125000   86.250000   
              wagon         94.500000  170.200000   69.000000   
porsche       convertible   89.500000  168.900000  207.000000   
              hardtop       89.500000  168.900000  207.000000   
              hatchback     98.400000  175.700000  288.000000   
toyota        hatchback     95.700000  158.700000   62.000000   
              wagon         97.900000  174.225000   85.500000   
volkswagen    sedan         97.300000  171.700000   72.250000   
volvo         sedan        104.300000  188.800000  114.000000   
              wagon        104.300000  188.800000  114.000000   

                           average-mileage         price  
company       body-style                                  
alfa-romero   convertible        21.000000  14997.500000  
              hatchback          19.000000  16500.000000  
audi          sedan              20.333333  15550.000000  
              wagon              19.000000  18920.000000  
bmw           sedan              19.000000  27213.333333  
chevrolet     hatchback          42.500000   5723.000000  
              sedan              38.000000   6575.000000  
dodge         hatchback          31.000000   6303.000000  
honda         sedan              24.500000  11645.000000  
              wagon              30.000000   7295.000000  
isuzu         sedan              33.333333   6785.000000  
jaguar        sedan              14.333333  34600.000000  
mazda         hatchback          27.250000   7482.500000  
              sedan              31.000000  18344.000000  
mercedes-benz hardtop            14.000000  45400.000000  
              sedan              18.000000  33256.000000  
              wagon              22.000000  28248.000000  
mitsubishi    hatchback          34.000000   5789.000000  
              sedan              25.000000   7589.000000  
nissan        sedan              31.500000   8524.000000  
              wagon              31.000000   7349.000000  
porsche       convertible        17.000000  37028.000000  
              hardtop            17.000000  34028.000000  
              hatchback          17.000000           NaN  
toyota        hatchback          32.333333   6058.000000  
              wagon              26.000000   9836.000000  
volkswagen    sedan              31.750000   8435.000000  
volvo         sedan              23.000000  12940.000000  
              wagon              23.000000  13415.000000

Jadvaldan xulosa qilsak, `volvo` firmasinig `sedan` turidagi mashinalarining o'rtacha qiymati 12940$, `wagon` turidagi mashinalarning o'rtacha narxi esa 13415$.

`.groupby()` yordamida nafaqat o'rta balki boshqa qiymatlarni ham hisoblash mumkin:

In [12]:
df.groupby(['company']).max()

,body-style,wheel-base,length,engine-type,num-of-cylinders,horsepower,average-mileage,price
company,,,,,,,,
alfa-romero,hatchback,94.5,171.2,ohcv,six,154,21,16500.0
audi,wagon,105.8,192.7,ohc,four,115,24,18920.0
bmw,sedan,110.0,197.0,ohc,six,182,23,41315.0
chevrolet,sedan,94.5,158.8,ohc,three,70,47,6575.0
dodge,hatchback,93.7,157.3,ohc,four,68,31,6377.0
honda,wagon,96.5,175.4,ohc,four,101,30,12945.0
isuzu,sedan,94.5,170.7,ohc,four,78,38,6785.0
jaguar,sedan,113.0,199.6,ohcv,twelve,262,15,36000.0
mazda,sedan,104.9,175.0,rotor,two,101,31,18344.0


`.agg()` metodi yordamida esa birdaniga bir nechta statistik ma'lumotlari olishimiz ham mumkin:

In [13]:
df.groupby('company').agg([len, min, max]) #len - guruhdagi elementlar soni, min - guruh uchun minimum, max - guruh uchun maximum

/tmp/ipykernel_17312/1300435525.py:1: FutureWarning: The provided callable <built-in function min> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  df.groupby('company').agg([len, min, max]) #len - guruhdagi elementlar soni, min - guruh uchun minimum, max - guruh uchun maximum
/tmp/ipykernel_17312/1300435525.py:1: FutureWarning: The provided callable <built-in function max> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df.groupby('company').agg([len, min, max]) #len - guruhdagi elementlar soni, min - guruh uchun minimum, max - guruh uchun maximum


body-style                         wheel-base                \
                     len          min        max        len    min    max   
company                                                                     
alfa-romero            3  convertible  hatchback          3   88.6   94.5   
audi                   4        sedan      wagon          4   99.4  105.8   
bmw                    6        sedan      sedan          6  101.2  110.0   
chevrolet              3    hatchback      sedan          3   88.4   94.5   
dodge                  2    hatchback  hatchback          2   93.7   93.7   
honda                  3        sedan      wagon          3   96.5   96.5   
isuzu                  3        sedan      sedan          3   94.3   94.5   
jaguar                 3        sedan      sedan          3  102.0  113.0   
mazda                  5    hatchback      sedan          5   93.1  104.9   
mercedes-benz          4      hardtop      wagon          4  110.0  120.9   
mitsubishi             4    hatchback      sedan          4   93.7   96.3   
nissan                 5        sedan      wagon          5   94.5  100.4   
porsche                3  convertible  hatchback          3   89.5   98.4   
toyota                 7    hatchback      wagon          7   95.7  104.5   
volkswagen             4        sedan      sedan          4   97.3   97.3   
volvo                  2        sedan      wagon          2  104.3  104.3   

              length               engine-type  ... num-of-cylinders  \
                 len    min    max         len  ...              max   
company                                         ...                    
alfa-romero        3  168.8  171.2           3  ...              six   
audi               4  176.6  192.7           4  ...             four   
bmw                6  176.8  197.0           6  ...              six   
chevrolet          3  141.1  158.8           3  ...            three   
dodge              2  157.3  157.3           2  ...             four   
honda              3  157.1  175.4           3  ...             four   
isuzu              3  155.9  170.7           3  ...             four   
jaguar             3  191.7  199.6           3  ...           twelve   
mazda              5  159.1  175.0           5  ...              two   
mercedes-benz      4  190.9  208.1           4  ...             five   
mitsubishi         4  157.3  172.4           4  ...             four   
nissan             5  165.3  184.6           5  ...              six   
porsche            3  168.9  175.7           3  ...              six   
toyota             7  158.7  187.8           7  ...              six   
volkswagen         4  171.7  171.7           4  ...             four   
volvo              2  188.8  188.8           2  ...             four   

              horsepower           average-mileage         price           \
                     len  min  max             len min max   len      min   
company                                                                     
alfa-romero            3  111  154               3  19  21     3  13495.0   
audi                   4  102  115               4  18  24     4  13950.0   
bmw                    6  101  182               6  15  23     6  16430.0   
chevrolet              3   48   70               3  38  47     3   5151.0   
dodge                  2   68   68               2  31  31     2   6229.0   
honda                  3   76  101               3  24  30     3   7295.0   
isuzu                  3   70   78               3  24  38     3   6785.0   
jaguar                 3  176  262               3  13  15     3  32250.0   
mazda                  5   68  101               5  17  31     5   5195.0   
mercedes-benz          4  123  184               4  14  22     4  25552.0   
mitsubishi             4   68   88               4  25  37     4   5389.0   
nissan                 5   55  152               5  19  45     5   6649.0   
porsche                3  

`.agg()` yordamida istalgan funksiyalarni chaqirishimiz mumkin, misol uchun `numpy` dagi `mean` (o'rta qiymat) funiksiyasini ham qo'shamiz. Natijalarni faqat `price` ustuni uchun chiqaramiz:

In [14]:
df.groupby('body-style').price.agg([len, min, max, np.mean])

/tmp/ipykernel_17312/614320205.py:1: FutureWarning: The provided callable <built-in function min> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  df.groupby('body-style').price.agg([len, min, max, np.mean])
/tmp/ipykernel_17312/614320205.py:1: FutureWarning: The provided callable <built-in function max> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df.groupby('body-style').price.agg([len, min, max, np.mean])
/tmp/ipykernel_17312/614320205.py:1: FutureWarning: The provided callable <function mean at 0x7a3d00921800> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df.groupby('body-style').price.agg([len, min, max, np.me

,len,min,max,mean
body-style,,,,
convertible,3,13495.0,37028.0,22341.000000
hardtop,2,34028.0,45400.0,39714.000000
hatchback,15,5151.0,16500.0,7159.571429
sedan,32,6575.0,41315.0,17706.333333
wagon,9,6918.0,28248.0,12730.111111


`.apply()` metodi yordamida guruhdagi qiymatlarni o'zgartirishimiz ham mumkin: